### PR TITLE
move punch scripts of ubuntu and sles to template

### PR DESCRIPTION
**Pull request by SharpRazor**
Signed-off-by: SharpRazor <bjcb@cn.ibm.com>

### CODE REVIEW

**Comment by SharpRazor** [comment]
the attach scripts generated on SLES vm is:
```
#!/bin/bash
# Generated by jinja2 template
fcp="1d11"
lun="0x0000000000000000"
target_filename="vdb"

echo "Enter SLES attach script with parameters: $fcp $lun $target_filename."

# see if zfcp is enable
enable_zfcp_mod=`lsmod | grep zfcp`
if [ -z "$enable_zfcp_mod" ];then
    modprobe zfcp
else
    echo "zfcp mode is already enabled"
fi

# online the fcp device
/sbin/cio_ignore -r $fcp > /dev/null
/sbin/chccwdev -e $fcp > /dev/null

declare -a ActiveWWPNs
ActiveWWPNs=(`ls /sys/bus/ccw/drivers/zfcp/0.0.$fcp/ | grep "0x"`)

# the host name of the FCP devices, for example
# the host name of "0.0.5c0c host0" is host0
# we need it to comprise the path of some files
HostName=`lszfcp | grep "$fcp" | cut -d " " -f 2`

# see if NPIV is enable or not
NPIV=`cat /sys/bus/ccw/drivers/zfcp/0.0.$fcp/$HostName/fc_host/$HostName/port_type`

# TODO: no auto port rescan?

# If auto-discovery of LUNs is disabled on s390 platforms
# luns need to be added to the configuration through
# the unit_add interface
AutoScan=`cat /sys/module/zfcp/parameters/allow_lun_scan`
if [[ "$AutoScan" != "Y" || "$NPIV" != "NPIV VPORT" ]]; then
    for wwpn in ${ActiveWWPNs[@]}
    do
        chzdev -e -a zfcp-lun 0.0.$fcp:$wwpn:$lun
        echo "$lun" > /sys/bus/ccw/drivers/zfcp/0.0.$fcp/$wwpn/unit_add
    done
fi


/sbin/zfcp_host_configure 0.0.$fcp 1

sleep 1
for wwpn in ${ActiveWWPNs[@]}
do
    /sbin/zfcp_disk_configure 0.0.$fcp $wwpn $lun 1
done

# create udev rules
sleep 1
touch /etc/udev/rules.d/51-zfcp-0.0.$fcp.rules
out=`cat "/etc/udev/rules.d/51-zfcp-0.0.$fcp.rules" | egrep -i "ccw/0.0.$fcp]online"`
if [[ ! $out ]]; then
  echo "ACTION==\"add\", SUBSYSTEM==\"ccw\", KERNEL==\"0.0.$fcp\", IMPORT{program}=\"collect 0.0.$fcp %k 0.0.$fcp zfcp\""| tee -a /etc/udev/rules.d/51-zfcp-0.0.$fcp.rules
  echo "ACTION==\"add\", SUBSYSTEM==\"drivers\", KERNEL==\"zfcp\", IMPORT{program}=\"collect 0.0.$fcp %k 0.0.$fcp zfcp\""| tee -a /etc/udev/rules.d/51-zfcp-0.0.$fcp.rules
  echo "ACTION==\"add\", ENV{COLLECT_0.0.$fcp}==\"0\", ATTR{[ccw/0.0.$fcp]online}=\"1\""| tee -a /etc/udev/rules.d/51-zfcp-0.0.$fcp.rules
fi
for wwpn in ${ActiveWWPNs[@]}
do
    echo "ACTION==\"add\", KERNEL==\"rport-*\", ATTR{port_name}==\"$wwpn\", SUBSYSTEMS==\"ccw\", KERNELS==\"0.0.$fcp\",ATTR{[ccw/0.0.$fcp]$wwpn/unit_add}=\"$lun\""| tee -a /etc/udev/rules.d/51-zfcp-0.0.$fcp.rules
done
if [[ $(which udevadm 2> /dev/null) != '' ]]; then
    udevadm settle
else
    udevsettle
fi

# send an iSCSI scan request with given host and optionally the ctl 
# the ctl means, c: channel,default to -
#                t: target, default to -
#                l: lun, default to -
all_hosts=(`ls /sys/class/scsi_host/`)
for host in ${all_hosts[@]}
do
    echo "- - -" > /sys/class/scsi_host/$host/scan
done

# wait for the devices ready
FoundDiskPath=0
for j in ${ActiveWWPNs[@]}
do
    x="/dev/disk/by-path/ccw-0.0.$fcp-zfcp-$j:$lun"
    # the x would be like:
    # ccw-0.0.1d13-zfcp-0x5005076306035388:0x4014400400000000
    echo "try disk path:$x"
    if [[ -e $x ]];then
        diskPath=$x
        FoundDiskPath=1
        echo "Found disk path is: $diskPath"
        break
    fi
done
if [ $FoundDiskPath -ne 1 ]; then
    echo "error happens during attachment because the device file not found, will exit with code 1."
    exit 1
fi

# ensure multipathd service is up
enable_multipath_mod=`lsmod | grep dm_multipath`
if [ -z "$enable_multipath_mod" ];then
    modprobe dm-multipath
    echo -e "#blacklist {
    #   devnode \"*\"
    #}
    " > /etc/multipath.conf
    mpathconf
    systemctl restart multipathd.service
else
    echo "dm-multipath mode is already enabled"
fi

WWID=`/lib/udev/scsi_id --page 0x83 --whitelisted $diskPath`
ConfigLib="/lib/udev/rules.d/56-zfcp.rules"
if [ -e "$ConfigLib" ]
then
    ConfigFile="/lib/udev/rules.d/56-zfcp.rules"
else
    ConfigFile="/etc/udev/rules.d/56-zfcp.rules"
fi

LinkItem="KERNEL==\"dm-*\", ENV{DM_UUID}==\"mpath-$WWID\", SYMLINK+=\"$target_filename\""
echo -e $LinkItem >> $ConfigFile

udevadm control --reload
udevadm trigger --sysname-match=dm-*

exit 0
```

**Comment by SharpRazor** [comment]
the detach script generated on SLES vm is:
```
#!/bin/bash
# Generated by jinja2 template
fcp="1d11"
lun="0x0000000000000000"
target_filename="vdb"

echo "Enter SLES detach script with parameters: $fcp $lun $target_filename."

# SourceDevice example: ccw-0.0.1c12-zfcp-0x5005076306035388:0x0001000000000000
SourceDevice=(`ls /dev/disk/by-path/ | grep "ccw-0.0.$fcp-zfcp-.*:$lun"`)

WWID=`/lib/udev/scsi_id --page 0x83 --whitelisted /dev/disk/by-path/$SourceDevice`

declare -a ActiveWWPNs
ActiveWWPNs=(`ls /sys/bus/ccw/drivers/zfcp/0.0.$fcp/ | grep "0x"`)

# exit code default to 0, because WWIDs may be empty
exit_code=0

# get the map name of the WWID, because multipath -f need map name
map_name=$(multipath -l $WWID -v 1)
echo "get map name: $map_name"
output=$(multipath -f $map_name 2>&1)
exit_code=$?

# error output not empty, means error happened
# and the error 'in use' and 'must provode a map name'
# of multipath -f will return same exit code 1
# so diff them, we will ingore the error of 'must provide a map name'
if [ "$output" ]; then
    if [ "$(echo $output | grep -i 'must provide a map name')" ]; then
        echo "ignore error on FCP $fcp and Lun $lun:$output"
        exit_code=0
    elif [ "$(echo $output | grep -i 'in use')" ]; then
        echo "FCP $fcp and Lun $lun is use:$output"
        exit_code=1
    else
        echo "unknown error on FCP $fcp and Lun $lun:$output"
        exit_code=2
    fi
fi

echo "exit code for multipath -f: $exit_code"
# if above code didn't succeed, exit now.
if [[ $exit_code != 0 ]]; then
    exit $exit_code
fi


RealPath=`readlink -f /dev/disk/by-path/$SourceDevice`
blockdev --flushbufs $RealPath > /dev/null
SdxName=`echo "$RealPath" | cut -d "/" -f 3`
echo 1 > /sys/block/$SdxName/device/delete

for WWPN in ${ActiveWWPNs[@]}
do
    /sbin/zfcp_disk_configure 0.0.$fcp $WWPN $lun 0
done

/sbin/zfcp_host_configure 0.0.$fcp 0

echo "traget file name is: $target_filename."
ConfigLib="/lib/udev/rules.d/56-zfcp.rules"
if [ -e "$ConfigLib" ]
then
    ConfigFile="/lib/udev/rules.d/56-zfcp.rules"
else
    ConfigFile="/etc/udev/rules.d/56-zfcp.rules"
fi
sed -i -e /SYMLINK+=\"$target_filename\"/d $ConfigFile

udevadm control --reload
udevadm trigger --sysname-match=dm-*

exit 0
```

**Comment by SharpRazor** [comment]
the attach script generated on UBUNTU vm is:
```
#!/bin/bash
# Generated by jinja2 template
fcp="1d14"
lun="0x0000000000000000"
target_filename="vdb"

echo "Enter UBUNTU atach script with parameters: $fcp $lun $target_filename."

# see if zfcp is enable
enable_zfcp_mod=`lsmod | grep zfcp`
if [ -z "$enable_zfcp_mod" ];then
    modprobe zfcp
else
    echo "zfcp mode is already enabled"
fi

declare -a ActiveWWPNs
ActiveWWPNs=(`ls /sys/bus/ccw/drivers/zfcp/0.0.$fcp/ | grep "0x"`)
HostName=`lszfcp | grep "$fcp" | cut -d " " -f 2`

/sbin/chzdev zfcp-host $fcp -e
for wwpn in ${ActiveWWPNs[@]}
do
    /sbin/chzdev zfcp-lun 0.0.$fcp:$wwpn:$lun -e
done

if [[ $(which udevadm 2> /dev/null) != '' ]]; then
    udevadm settle
else
    udevsettle
fi

# send an iSCSI scan request with given host and optionally the ctl 
# the ctl means, c: channel,default to -
#                t: target, default to -
#                l: lun, default to -
all_hosts=(`ls /sys/class/scsi_host/`)
for host in ${all_hosts[@]}
do
    echo "- - -" > /sys/class/scsi_host/$host/scan
done

# TODO:only ubuntu 20 apply this?

# wait for the devices ready
# when attach v7k volume:
# 1. if the lun id less than 256,
#    the file under /dev/disk/by-path/ will as below,
#    take 'lun id = 0' as example:
#    ccw-0.0.5c03-fc-0x5005076802400c1a-lun-0, the the lun id is decimal.
# 2. if the lun id is equal or more than 256,
#    the file under /dev/disk/by-path/ will as below,
#    take 'lun id = 256' as example:
#    ccw-0.0.1a0d-fc-0x500507680b26bac7-lun-0x0100000000000000,
#    the lun id is hex.

# when attach ds8k volume:
#    the file under /dev/disk/by-path/ will as below,
#    take "volume id 140c" as example:
#    ccw-0.0.1a0d-fc-0x5005076306035388-lun-0x4014400c00000000,
#    the lun id is always hex.

# take 0x0100000000000000 as example
# remove first 2 chars '0x' and return the next 4 chars '0100'
hexLun=${lun:2:4}
# transfer hex to decimal
lun_id=$(( 16#$hexLun ))
if [[ $lun_id -lt 256 ]];then
    target_lun=$lun_id
else
    target_lun=$lun
fi

# wait for the devices show up
FoundDiskPath=0
for j in ${ActiveWWPNs[@]}
do
    x="/dev/disk/by-path/ccw-0.0.$fcp-fc-.*-lun-$target_lun"
    echo "try disk path:$x"
    if [[ -e $x ]];then
        diskPath=$x
        FoundDiskPath=1
        echo "Found disk path is: $diskPath"
        break
    fi
done
if [ $FoundDiskPath -ne 1 ]; then
    echo "error happens during attachment because the device file not found, will exit with code 1."
    exit 1
fi

# ensure multipathd service is up
enable_multipath_mod=`lsmod | grep dm_multipath`
if [ -z "$enable_multipath_mod" ];then
    modprobe dm-multipath
    echo -e "#blacklist {
    #   devnode \"*\"
    #}
    " > /etc/multipath.conf
    mpathconf
    systemctl restart multipathd.service
else
    echo "dm-multipath mode is already enabled"
fi

WWID=`/lib/udev/scsi_id --page 0x83 --whitelisted $diskPath`
ConfigLib="/lib/udev/rules.d/56-zfcp.rules"
if [ -e "$ConfigLib" ]
then
    ConfigFile="/lib/udev/rules.d/56-zfcp.rules"
else
    ConfigFile="/etc/udev/rules.d/56-zfcp.rules"
fi

LinkItem="KERNEL==\"dm-*\", ENV{DM_UUID}==\"mpath-$WWID\", SYMLINK+=\"$target_filename\""
echo -e $LinkItem >> $ConfigFile

udevadm control --reload
udevadm trigger --sysname-match=dm-*

exit 0
```

**Comment by SharpRazor** [comment]
the detach script generated on UBUNTU vm is:
```
#!/bin/bash
# Generated by jinja2 template
fcp="1d14"
lun="0x0000000000000000"
target_filename="vdb"

echo "Enter UBUNTU detach script with parameters: $fcp $lun $target_filename."

# TODO:only ubuntu 20 apply this?

# wait for the devices ready
# when attach v7k volume:
# 1. if the lun id less than 256,
#    the file under /dev/disk/by-path/ will as below,
#    take 'lun id = 0' as example:
#    ccw-0.0.5c03-fc-0x5005076802400c1a-lun-0, the the lun id is decimal.
# 2. if the lun id is equal or more than 256,
#    the file under /dev/disk/by-path/ will as below,
#    take 'lun id = 256' as example:
#    ccw-0.0.1a0d-fc-0x500507680b26bac7-lun-0x0100000000000000,
#    the lun id is hex.

# when attach ds8k volume:
#    the file under /dev/disk/by-path/ will as below,
#    take "volume id 140c" as example:
#    ccw-0.0.1a0d-fc-0x5005076306035388-lun-0x4014400c00000000,
#    the lun id is always hex.

# take 0x0100000000000000 as example
# remove first 2 chars '0x' and return the next 4 chars '0100'
hexLun=${lun:2:4}
# transfer hex to decimal
lun_id=$(( 16#$hexLun ))
if [[ $lun_id -lt 256 ]];then
    target_lun=$lun_id
else
    target_lun=$lun
fi
SourceDevice=(`ls /dev/disk/by-path/ | grep "ccw-0.0.$fcp-fc-.*-lun-$target_lun"`)


WWID=`/lib/udev/scsi_id --page 0x83 --whitelisted /dev/disk/by-path/$SourceDevice`

declare -a ActiveWWPNs
ActiveWWPNs=(`ls /sys/bus/ccw/drivers/zfcp/0.0.$fcp/ | grep "0x"`)

# exit code default to 0, because WWID may be empty
exit_code=0

# get the map name of the WWID, because multipath -f need map name
map_name=$(multipath -l $WWID -v 1)
echo "get map name: $map_name"
output=$(multipath -f $map_name 2>&1)
exit_code=$?

# error output not empty, means error happened
# and the error 'in use' and 'must provode a map name'
# of multipath -f will return same exit code 1
# so diff them, we will ingore the error of 'must provide a map name'
if [ "$output" ]; then
    if [ "$(echo $output | grep -i 'must provide a map name')" ]; then
        echo "ignore error on FCP $fcp and Lun $lun:$output"
        exit_code=0
    elif [ "$(echo $output | grep -i 'in use')" ]; then
        echo "FCP $fcp and Lun $lun is use:$output"
        exit_code=1
    else
        echo "unknown error on FCP $fcp and Lun $lun:$output"
        exit_code=2
    fi
fi

echo "exit code for multipath -f: $exit_code"
# if above code didn't succeed, exit now.
if [[ $exit_code != 0 ]]; then
    exit $exit_code
fi

RealPath=`readlink -f /dev/disk/by-path/$SourceDevice`
blockdev --flushbufs $RealPath > /dev/null
SdxName=`echo "$RealPath" | cut -d "/" -f 3`
echo 1 > /sys/block/$SdxName/device/delete


for WWPN in ${ActiveWWPNs[@]}
do
    /sbin/chzdev zfcp-lun 0.0.$fcp:$WWPN:$lun -d
done
/sbin/chzdev zfcp-host $fcp -d

ConfigLib="/lib/udev/rules.d/56-zfcp.rules"
if [ -e "$ConfigLib" ]
then
    ConfigFile="/lib/udev/rules.d/56-zfcp.rules"
else
    ConfigFile="/etc/udev/rules.d/56-zfcp.rules"
fi
sed -i -e /SYMLINK+=\"$target_filename\"/d $ConfigFile

udevadm control --reload
udevadm trigger --sysname-match=dm-*

exit 0
```

**Review by dongyanyang — APPROVED** [review]
lgtm

**Comment by dongyanyang** [comment]
Wait for @SharpRazor to do more test.
@SharpRazor please attach the cases you have done after all tests passed. Then @bjhuangr and @jichenjc can help to review and merge. Thanks!

**Comment by SharpRazor** [comment]
Test cases:
## UBUNTU VM
### 0, deploy a VM, target VM's power status is Active
### 1. attach 3 SVC and 3 DS8K volumes to target VM ==> pass
### 2. detach 1 SVC volume and 1 DS8K volume from it.
### 3. reboot target VM and check the volume still exist or not
### 4. detach all the rest volumes, to see the FCP devices offline or not
### 5. shutdown the target VM
### 6. attach 3 SVC and 3 DS8K volumes to target VM
### 7. detach 1 SVC volume and 1 DS8K volume from it.
### 8. start the target VM to check the status volumes
### 9. shutdown the VM again
### 10. detach all the rest volumes
### 11. start the target VM to check the status of volumes and FCP devices

## SLES VM
### 0, deploy a VM, target VM's power status is Active
### 1. attach 3 SVC and 3 DS8K volumes to target VM ==> pass
### 2. detach 1 SVC volume and 1 DS8K volume from it.
### 3. reboot target VM and check the volume still exist or not
### 4. detach all the rest volumes, to see the FCP devices offline or not
### 5. shutdown the target VM
### 6. attach 3 SVC and 3 DS8K volumes to target VM
### 7. detach 1 SVC volume and 1 DS8K volume from it.
### 8. start the target VM to check the status volumes
### 9. shutdown the VM again
### 10. detach all the rest volumes
### 11. start the target VM to check the status of volumes and FCP devices

**Comment by SharpRazor** [comment]
## SLES VM
### 0. deploy a VM, target VM's power status is Active
### 1. attach 3 SVC and 3 DS8K volumes to target VM
### 2. detach 1 SVC volume and 1 DS8K volume from it.
### 3. reboot target VM and check the volume still exist or not
### 4. detach all the rest volumes, to see the FCP devices offline or not
### 5. shutdown the target VM
### 6. attach 3 SVC and 3 DS8K volumes to target VM
### 7. detach 1 SVC volume and 1 DS8K volume from it.
### 8. start the target VM to check the status volumes
### 9. shutdown the VM again
### 10. detach all the rest volumes
### 11. start the target VM to check the status of volumes and FCP devices

**Comment by SharpRazor** [comment]
## UBUNTU VM2
### 0. deploy another ubuntu VM ==> pass
### 1. attach 4 SVC and 2 DS8K volumes ==> pass
```
dummyuser@ubuntu20-2:~$ lszfcp
0.0.1a11 host0
0.0.1b13 host1
0.0.1c13 host2
0.0.1d10 host3
dummyuser@ubuntu20-2:~$ sudo multipath -l
[sudo] password for dummyuser:
mpathe (3600507640083826de000000000004793) dm-4 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:10:2          sdab 65:176 active undef running
| |- 1:0:4:2           sdad 65:208 active undef running
| |- 0:0:8:2           sdz  65:144 active undef running
| |- 2:0:5:2           sdas 66:192 active undef running
| |- 2:0:7:2           sdaw 67:0   active undef running
| |- 3:0:9:2           sdby 68:192 active undef running
| `- 3:0:3:2           sdbm 68:0   active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:9:2           sdaa 65:160 active undef running
  |- 1:0:0:2           sdac 65:192 active undef running
  |- 0:0:6:2           sdy  65:128 active undef running
  |- 2:0:3:2           sdao 66:128 active undef running
  |- 2:0:8:2           sdba 67:64  active undef running
  |- 3:0:5:2           sdbq 68:64  active undef running
  `- 3:0:7:2           sdbu 68:128 active undef running
mpathd (3600507640083826de000000000004791) dm-3 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:6:1           sdh  8:112  active undef running
| |- 0:0:9:1           sdl  8:176  active undef running
| |- 1:0:0:1           sdp  8:240  active undef running
| |- 2:0:3:1           sdan 66:112 active undef running
| |- 2:0:8:1           sdaz 67:48  active undef running
| |- 3:0:5:1           sdbp 68:48  active undef running
| `- 3:0:7:1           sdbt 68:112 active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:8:1           sdj  8:144  active undef running
  |- 0:0:10:1          sdn  8:208  active undef running
  |- 1:0:4:1           sdx  65:112 active undef running
  |- 2:0:7:1           sdav 66:240 active undef running
  |- 2:0:5:1           sdar 66:176 active undef running
  |- 3:0:3:1           sdbl 67:240 active undef running
  `- 3:0:9:1           sdbx 68:176 active undef running
mpathc (3600507640083826de000000000004794) dm-2 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:9:0           sdk  8:160  active undef running
| |- 0:0:6:0           sdg  8:96   active undef running
| |- 1:0:0:0           sdo  8:224  active undef running
| |- 2:0:3:0           sdam 66:96  active undef running
| |- 2:0:8:0           sday 67:32  active undef running
| |- 3:0:5:0           sdbo 68:32  active undef running
| `- 3:0:7:0           sdbs 68:96  active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:10:0          sdm  8:192  active undef running
  |- 0:0:8:0           sdi  8:128  active undef running
  |- 1:0:4:0           sdw  65:96  active undef running
  |- 2:0:5:0           sdaq 66:160 active undef running
  |- 2:0:7:0           sdau 66:224 active undef running
  |- 3:0:3:0           sdbk 67:224 active undef running
  `- 3:0:9:0           sdbw 68:160 active undef running
mpathb (36005076306ffd3880000000000001538) dm-1 IBM,2107900
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
`-+- policy='service-time 0' prio=0 status=active
  |- 0:0:0:1077428245  sdb  8:16   active undef running
  |- 0:0:1:1077428245  sdd  8:48   active undef running
  |- 0:0:2:1077428245  sdf  8:80   active undef running
  |- 1:0:2:1077428245  sdt  65:48  active undef running
  |- 1:0:3:1077428245  sdv  65:80  active undef running
  |- 1:0:1:1077428245  sdr  65:16  active undef running
  |- 2:0:0:1077428245  sdal 66:80  active undef running
  |- 2:0:9:1077428245  sdbd 67:112 active undef running
  |- 2:0:10:1077428245 sdbf 67:144 active undef running
  |- 3:0:1:1077428245  sdbj 67:208 active undef running
  |- 3:0:0:1077428245  sdbh 67:176 active undef running
  `- 3:0:10:1077428245 sdcb 68:240 active undef running
mpatha (36005076306ffd3880000000000001508) dm-0 IBM,2107900
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
`-+- policy='service-time 0' prio=0 status=active
  |- 0:0:1:1074282517  sdc  8:32   active undef running
  |- 0:0:0:1074282517  sda  8:0    active undef running
  |- 0:0:2:1074282517  sde  8:64   active undef running
  |- 1:0:1:1074282517  sdq  65:0   active undef running
  |- 1:0:2:1074282517  sds  65:32  active undef running
  |- 1:0:3:1074282517  sdu  65:64  active undef running
  |- 2:0:0:1074282517  sdak 66:64  active undef running
  |- 2:0:9:1074282517  sdbc 67:96  active undef running
  |- 2:0:10:1074282517 sdbe 67:128 active undef running
  |- 3:0:0:1074282517  sdbg 67:160 active undef running
  |- 3:0:1:1074282517  sdbi 67:192 active undef running
  `- 3:0:10:1074282517 sdca 68:224 active undef running
mpathf (3600507640083826de000000000004792) dm-5 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:6:3           sdae 65:224 active undef running
| |- 1:0:0:3           sdai 66:32  active undef running
| |- 0:0:9:3           sdag 66:0   active undef running
| |- 2:0:3:3           sdap 66:144 active undef running
| |- 2:0:8:3           sdbb 67:80  active undef running
| |- 3:0:5:3           sdbr 68:80  active undef running
| `- 3:0:7:3           sdbv 68:144 active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:8:3           sdaf 65:240 active undef running
  |- 0:0:10:3          sdah 66:16  active undef running
  |- 1:0:4:3           sdaj 66:48  active undef running
  |- 2:0:7:3           sdax 67:16  active undef running
  |- 2:0:5:3           sdat 66:208 active undef running
  |- 3:0:3:3           sdbn 68:16  active undef running
  `- 3:0:9:3           sdbz 68:208 active undef running
dummyuser@ubuntu20-2:~$
```

### 3. detach 1 SVC and 1 DS8K volume
```
dummyuser@ubuntu20-2:~$ sudo multipath -l
mpathe (3600507640083826de000000000004793) dm-4 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:10:2          sdab 65:176 active undef running
| |- 1:0:4:2           sdad 65:208 active undef running
| |- 0:0:8:2           sdz  65:144 active undef running
| |- 2:0:5:2           sdas 66:192 active undef running
| |- 2:0:7:2           sdaw 67:0   active undef running
| |- 3:0:9:2           sdby 68:192 active undef running
| `- 3:0:3:2           sdbm 68:0   active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:9:2           sdaa 65:160 active undef running
  |- 1:0:0:2           sdac 65:192 active undef running
  |- 0:0:6:2           sdy  65:128 active undef running
  |- 2:0:3:2           sdao 66:128 active undef running
  |- 2:0:8:2           sdba 67:64  active undef running
  |- 3:0:5:2           sdbq 68:64  active undef running
  `- 3:0:7:2           sdbu 68:128 active undef running
mpathc (3600507640083826de000000000004794) dm-2 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:9:0           sdk  8:160  active undef running
| |- 0:0:6:0           sdg  8:96   active undef running
| |- 1:0:0:0           sdo  8:224  active undef running
| |- 2:0:3:0           sdam 66:96  active undef running
| |- 2:0:8:0           sday 67:32  active undef running
| |- 3:0:5:0           sdbo 68:32  active undef running
| `- 3:0:7:0           sdbs 68:96  active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:10:0          sdm  8:192  active undef running
  |- 0:0:8:0           sdi  8:128  active undef running
  |- 1:0:4:0           sdw  65:96  active undef running
  |- 2:0:5:0           sdaq 66:160 active undef running
  |- 2:0:7:0           sdau 66:224 active undef running
  |- 3:0:3:0           sdbk 67:224 active undef running
  `- 3:0:9:0           sdbw 68:160 active undef running
mpatha (36005076306ffd3880000000000001508) dm-0 IBM,2107900
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
`-+- policy='service-time 0' prio=0 status=active
  |- 0:0:1:1074282517  sdc  8:32   active undef running
  |- 0:0:0:1074282517  sda  8:0    active undef running
  |- 0:0:2:1074282517  sde  8:64   active undef running
  |- 1:0:1:1074282517  sdq  65:0   active undef running
  |- 1:0:2:1074282517  sds  65:32  active undef running
  |- 1:0:3:1074282517  sdu  65:64  active undef running
  |- 2:0:0:1074282517  sdak 66:64  active undef running
  |- 2:0:9:1074282517  sdbc 67:96  active undef running
  |- 2:0:10:1074282517 sdbe 67:128 active undef running
  |- 3:0:0:1074282517  sdbg 67:160 active undef running
  |- 3:0:1:1074282517  sdbi 67:192 active undef running
  `- 3:0:10:1074282517 sdca 68:224 active undef running
mpathf (3600507640083826de000000000004792) dm-5 IBM,2145
size=1.0G features='1 queue_if_no_path' hwhandler='1 alua' wp=rw
|-+- policy='service-time 0' prio=0 status=active
| |- 0:0:6:3           sdae 65:224 active undef running
| |- 1:0:0:3           sdai 66:32  active undef running
| |- 0:0:9:3           sdag 66:0   active undef running
| |- 2:0:3:3           sdap 66:144 active undef running
| |- 2:0:8:3           sdbb 67:80  active undef running
| |- 3:0:5:3           sdbr 68:80  active undef running
| `- 3:0:7:3           sdbv 68:144 active undef running
`-+- policy='service-time 0' prio=0 status=enabled
  |- 0:0:8:3           sdaf 65:240 active undef running
  |- 0:0:10:3          sdah 66:16  active undef running
  |- 1:0:4:3           sdaj 66:48  active undef running
  |- 2:0:7:3           sdax 67:16  active undef running
  |- 2:0:5:3           sdat 66:208 active undef running
  |- 3:0:3:3           sdbn 68:16  active undef running
  `- 3:0:9:3           sdbz 68:208 active undef running
dummyuser@ubuntu20-2:~$
```

### 4 detach all the rest volumes


### 5. attach 2 SVC and 1 DS8K volumes again


### 6. reboot and check

**Comment by SharpRazor** [comment]
## SLES VM2

### 0. deploy another sles VM
### 1. attach 4 SVC and 2 DS8K volumes
### 2. detach 1 svc and 1 DS8K volumes
### 3. detach all the rest volumes

### 4. attach 2 SVC and 1 DS8K volumes

### 5. reboot and check

File: zvmsdk/volumeops/templates/sles_detach_volume.j2 [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        exit_code=1
+    else
+        echo "unknown error on FCP $fcp and Lun $lun:$output"
+        exit_code=2
+    fi
+fi
+
+echo "exit code for multipath -f: $exit_code"
+# if above code didn't succeed, exit now.
+if [[ $exit_code != 0 ]]; then
```
Comment (dongyanyang) [inline zvmsdk/volumeops/templates/sles_detach_volume.j2:47]: != -> -ne

File: zvmsdk/volumeops/templates/sles_detach_volume.j2 [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    exit $exit_code
+fi
+
+RealPath=`readlink -f /dev/disk/by-path/$SourceDevice`
+blockdev --flushbufs $RealPath > /dev/null
+SdxName=`echo "$RealPath" | cut -d "/" -f 3`
+echo 1 > /sys/block/$SdxName/device/delete
```
Comment (dongyanyang) [inline zvmsdk/volumeops/templates/sles_detach_volume.j2:54]: why would do this 'delete' for this only single path sdx?

File: zvmsdk/volumeops/templates/ubuntu_detach_volume.j2 [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        exit_code=1
+    else
+        echo "unknown error on FCP $fcp and Lun $lun:$output"
+        exit_code=2
+    fi
+fi
+
+echo "exit code for multipath -f: $exit_code"
+# if above code didn't succeed, exit now.
+if [[ $exit_code != 0 ]]; then
```
Comment (dongyanyang) [inline zvmsdk/volumeops/templates/ubuntu_detach_volume.j2:57]: !=  to -ne

File: zvmsdk/volumeops/templates/ubuntu_detach_volume.j2 [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    exit $exit_code
+fi
+
+RealPath=`readlink -f /dev/disk/by-path/$SourceDevice`
+blockdev --flushbufs $RealPath > /dev/null
+SdxName=`echo "$RealPath" | cut -d "/" -f 3`
```
Comment (dongyanyang) [inline zvmsdk/volumeops/templates/ubuntu_detach_volume.j2:63]: same comments as sles

File: zvmsdk/volumeops/templates/sles_detach_volume.j2 (hunk 2 above)
Comment (SharpRazor) [reply zvmsdk/volumeops/templates/sles_detach_volume.j2:54]: this is history reason, sometimes may found leftovers under `multipath -l`, this is trying to delete the leftovers.
I can remove them in future PR.

File: zvmsdk/volumeops/templates/ubuntu_detach_volume.j2 (hunk 3 above)
Comment (SharpRazor) [reply zvmsdk/volumeops/templates/ubuntu_detach_volume.j2]: done.

File: zvmsdk/volumeops/templates/sles_detach_volume.j2 (hunk 1 above)
Comment (SharpRazor) [reply zvmsdk/volumeops/templates/sles_detach_volume.j2]: done.

File: zvmsdk/volumeops/templates/sles_detach_volume.j2 (hunk 2 above)
Comment (dongyanyang) [reply zvmsdk/volumeops/templates/sles_detach_volume.j2:54]: please open issue to track the detail of leftover issue.

**Review by dongyanyang — APPROVED** [review]
lgtm. The dead code of Ubuntu and SLES in dist.py will be deleted with future PR.
@jichenjc @bjhuangr please review.

**Pull request merged**